offBy1_checker.go

```diff
@@ -0,0 +1,66 @@
+package checkers
+
+import (
+	"go/ast"
+	"go/token"
+
+	"github.com/go-lintpack/lintpack"
+	"github.com/go-lintpack/lintpack/astwalk"
+	"github.com/go-toolsmith/astcast"
+	"github.com/go-toolsmith/astcopy"
+	"github.com/go-toolsmith/astequal"
+	"github.com/go-toolsmith/typep"
+)
+
+func init() {
+	var info lintpack.CheckerInfo
+	info.Name = "offBy1"
+	info.Tags = []string{"diagnostic", "experimental"}
+	info.Summary = "Detects various off-by-one kind of errors"
+	info.Before = `xs[len(xs)]`
+	info.After = `xs[len(xs)-1]`
+
+	collection.AddChecker(&info, func(ctx *lintpack.CheckerContext) lintpack.FileWalker {
+		return astwalk.WalkerForExpr(&offBy1Checker{ctx: ctx})
+	})
+}
+
+type offBy1Checker struct {
+	astwalk.WalkHandler
+	ctx *lintpack.CheckerContext
+}
+
+func (c *offBy1Checker) VisitExpr(e ast.Expr) {
+	// TODO(Quasilyte): handle more off-by-1 patterns.
+	// TODO(Quasilyte): check whether go/analysis can help here.
+
+	// Detect s[len(s)] expressions that always panic.
+	// The correct form is s[len(s)-1].
```

Comment:
Do we need this comment?

Reply:
You can remove it by moving function body to a separate method, so when there are multiple patterns that are matched, it's apparent which method matches which code.

Reply:
got it